cfg/rippled-example.cfg

```diff
@@ -148,7 +148,8 @@
 #   ip = <IP-address>
 #
 #       Required. Determines the IP address of the network interface to bind
-#       to. To bind to all available interfaces, uses 0.0.0.0
+#       to. To bind to all available interfaces, use 0.0.0.0
+#       For binding to both IPv4 and IPv6 interfaces, use ::
```

Comment:
I think that this comment doesn't account for differences between platforms. Windows, according to https://serverfault.com/questions/21657/semantics-of-and-0-0-0-0-in-dual-stack-oses,  will bind only to IPv6 addresses when `::` is used.

(@ximinez, @miguelportilla: if you have IPv6 enabled & available can you check on Windows? @HowardHinnant, @scottschurr: if you have IPv6 enabled & available can you check on Mac?)

Would it be prudent to link to the above answer and reword this comment slightly:

>The special notation 0.0.0.0 will bind on all available IPv4 interfaces. The special notation :: will bind to all available IPv6 interfaces and may, depending on your operating system and configuration, also bind on all IPv4 interfaces.
>For more, please consult https://serverfault.com/questions/21657/semantics-of-and-0-0-0-0-in-dual-stack-oses and your O/S documentation.

Reply:
Under Windows, my current config specifies `ip = 127.0.0.1` and `admin = 127.0.0.1`.
Test 1. Control: I started in standalone with the current config.
```
$ rippled.exe --rpc_ip=127.0.0.1:5015 -q ping`
{
   "result" : {
      "role" : "admin",
      "status" : "success"
   }
}
$ rippled.exe --rpc_ip=[::1]:5015 -q ping           {
   "error" : "internal",
   "error_code" : 71,
   "error_message" : "Internal error.",
   "error_what" : "no response from server"
}
# IPv6 address reported by ipconfig
$ ./build/cmake/msvc.ON/Debug/rippled.exe --rpc_ip=[2600:8807:c140:8ef:5117:598a:eeba:3633]:5015 -q ping
{
   "error" : "internal",
   "error_code" : 71,
   "error_message" : "Internal error.",
   "error_what" : "no response from server"
}
# Link-local IPv6 reported by ipconfig
$ ./build/cmake/msvc.ON/Debug/rippled.exe --rpc_ip=[fe80::5117:598a:eeba:3633]:5015 -q ping
{
   "error" : "internal",
   "error_code" : 71,
   "error_message" : "Internal error.",
   "error_what" : "no response from server"
}
```
Test 2. All IPv4 (`ip = 0.0.0.0`). 
```
$ rippled.exe --rpc_ip=127.0.0.1:5015 -q ping
{
   "result" : {
      "role" : "admin",
      "status" : "success"
   }
}
$ rippled.exe --rpc_ip=[::1]:5015 -q ping           {
   "error" : "internal",
   "error_code" : 71,
   "error_message" : "Internal error.",
   "error_what" : "no response from server"
}
# IPv6 address reported by ipconfig
$ ./build/cmake/msvc.ON/Debug/rippled.exe --rpc_ip=[2600:8807:c140:8ef:5117:598a:eeba:3633]:5015 -q ping
{
   "error" : "internal",
   "error_code" : 71,
   "error_message" : "Internal error.",
   "error_what" : "no response from server"
}
# Link-local IPv6 reported by ipconfig
$ ./build/cmake/msvc.ON/Debug/rippled.exe --rpc_ip=[fe80::5117:598a:eeba:3633]:5015 -q ping
{
   "error" : "internal",
   "error_code" : 71,
   "error_message" : "Internal error.",
   "error_what" : "no response from server"
}
```
Test 3: All IPv6 (`ip = ::`)
```
$ rippled.exe --rpc_ip=127.0.0.1:5015 -q ping
{
   "result" : {
      "status" : "success"
   }
}

$ /rippled.exe --rpc_ip=[::1]:5015 -q ping
{
   "result" : {
      "status" : "success"
   }
}
$ rippled.exe --rpc_ip=[2600:8807:c140:8ef:5117:598a:eeba:3633]:5015 -q ping
{
   "result" : {
      "status" : "success"
   }
}
$ rippled.exe --rpc_ip=[fe80::5117:598a:eeba:3633]:5015 -q ping
{
   "result" : {
      "status" : "success"
   }
}
```
*Notice that none of those options report an admin connection. This is pretty disappointing.*
Test 4: All IPv6 with IPv6 admin (`ip = ::`, `admin = ::1`)
```
$ rippled.exe --rpc_ip=127.0.0.1:5015 -q ping
{
   "result" : {
      "status" : "success"
   }
}
$ rippled.exe --rpc_ip=[::1]:5015 -q ping
{
   "result" : {
      "role" : "admin",
      "status" : "success"
   }
}
$ rippled.exe --rpc_ip=[2600:8807:c140:8ef:5117:598a:eeba:3633]:5015 -q ping
{
   "result" : {
      "status" : "success"
   }
}
$ rippled.exe --rpc_ip=[fe80::5117:598a:eeba:3633]:5015 -q ping
{
   "result" : {
      "status" : "success"
   }
}
```
Conclusion:
It looks like `::` under Windows causes ripple to listen on all IPv4 and IPv6 addresses, but does not allow an IPv4 address to be the admin connection. If you want IPv6, you'll have to configure an IPv6 admin address.

Reply:
After reading the serverfault article, I checked what netstat reports with rippled running listening to `::`. It shows us listening to both IPv4 and IPv6:
```
$ netstat -an | grep "5015"
  TCP    0.0.0.0:5015           0.0.0.0:0              LISTENING
  TCP    [::]:5015              [::]:0                 LISTENING
```
So either rippled has a special case or Windows has some automatic workaround.

Reply:
I think it might be asio on windows setting this option (https://github.com/boostorg/asio/blob/25dc6780c2c73dd6a4d74e65e854fc0f705cbb60/include/boost/asio/detail/impl/socket_ops.ipp#L1394-L1408), although I might be reading that wrong.

Reply:
FWIW, I poked at this a bit on macOS with @mellery451's help.  On a stand alone instance, if I set the following in my config file:
```
[port_rpc]
ip = ::
port = 5005
protocol = http
admin = ::
```
then I get the following responses to the next two commands:
```
$ build/rippled --rpc_ip=127.0.0.1:5005 -q ping
{
   "result" : {
      "role" : "admin",
      "status" : "success"
   }
}
```
```
$ build/rippled --rpc_ip=[::]:5005 -q ping
{
   "result" : {
      "role" : "admin",
      "status" : "success"
   }
}
```